File: libgomp/testsuite/libgomp.c-c++-common/taskloop-1.c
/* { dg-do run } */
/* { dg-options "-O2" } */
/* { dg-additional-options "-std=c99" { target c } } */

int q, r, e;

__attribute__((noinline, noclone)) void
foo (long a, long b)
{
  #pragma omp taskloop lastprivate (q) nogroup
    for (long d = a; d < b; d += 2)
      {
	q = d;
	if (d < 2 || d > 6 || (d & 1))
	  #pragma omp atomic
	    e |= 1;
      }
}

__attribute__((noinline, noclone)) int
bar (int a, int b)
{
  int q = 7;
  #pragma omp taskloop lastprivate (q)
    for (int d = a; d < b; d++)
      {
	if (d < 12 || d > 17)
	  #pragma omp atomic
	    e |= 1;
	q = d;
      }
  return q;
}

int
main ()
{
  #pragma omp parallel
    #pragma omp single
      {
	foo (2, 7);
	r = bar (12, 18);
      }
  if (q != 6 || r != 17 || e)
    __builtin_abort ();
  return 0;
}

File: libgomp/testsuite/libgomp.fortran/taskloop1.f90
! { dg-do run }
  common /blk/ q, e
  integer :: q, r
  logical :: e
!$omp parallel
!$omp single
  call foo (2, 7)
  r = bar (12, 18)
!$omp end single
!$omp end parallel
  if (q .ne. 6 .or. r .ne. 17 .or. e) stop 1
contains
  subroutine foo (a, b)
    integer, intent (in) :: a, b
    common /blk/ q, e
    integer :: q, r, d
    logical :: e
!$omp taskloop lastprivate (q) nogroup
    do d = a, b, 2
      q = d
      if (d < 2 .or. d > 6 .or. iand (d, 1) .ne. 0) then
!$omp atomic write
        e = .true.
      end if
    end do
  end subroutine foo
  function bar (a, b)
    integer, intent (in) :: a, b
    integer :: bar
    common /blk/ q, e
    integer :: q, r, d, s
    logical :: e
    s = 7
!$omp taskloop lastprivate (s)
    do d = a, b - 1
      if (d < 12 .or. d > 17) then
!$omp atomic write
        e = .true.
      end if
      s = d
    end do
!$omp end taskloop
    bar = s
  end function bar
end
